### PR TITLE
vehicle speed rework

### LINKED ISSUE #25652: Player friendly vehicle speed rework
CDDA currently calculates vehicle speeds in a way that is not really connected to reality. Light weight, low powered vehicles are too slow, heavy vehicles are too fast, and the maximum speed of a vehicle is based solely on engine power and is much much too high for large engines.

Change the formulas to better relate to reality.

**Describe the solution you'd like**  
Basically implement a somewhat simplified version of this: http://www.asawicki.info/Mirror/Car%20Physics%20for%20Games/Car%20Physics%20for%20Games.html

We keep the current system of calculating engine power. 
However, instead of calculating drag deceleration as:
```
slowdown (mph/turn) = 100 / ( 2 + 9 * k_dynamics * k_mass )
    where k_dynamics = k_aero * k_friction
        (k_areo is a number between 0.6 and .9, calculated roughly on body shape)
        (k_friction is a number between 0.6 an 0.98, calculated as 9000 / ( 9000 + wheel area ) )
    and k_mass = 50 / ( 50 + vehicle_mass ) / ( 0.88 * wheel area ) - basically 0.9 to 0.98
```

we calculate slowdown based on air resistance and rolling resistance.
ie,
```
slowdown ( m/s^2 ) = ( Cdair * speed^2 + Crolling * speed ) / vehicle mass
where 
    Cdair = .5 * air density (1.29 kg/m^3) * aero drag coefficient * vehicle front area
        where drag coefficient is a number between 0.3 and 1, depending on vehicle 
        shape (we could use k_aero for now)
    vehicle front area is proportional to vehicle width (roughly 0.5m / tile) * highest board 
    height (say, 1.4m for half boards, 3m for full boards)
    Crolling = vehicle mass * number of tires * tire C (roughly 0.002 for small tires to 
    0.0075 for armored tires)
```
Similarly, instead of calculating safe velocity, max velocity, and acceleration as
```
safe velocity = engine power * m2c * k_dynamics * k_mass * 80
max velocity = engine power * 80
acceleration = safe velocity * k_mass / 20
```
we calculate
```
acceleration = engine power / current speed / mass
    ( this looks a little weird, but acceleration is F at the wheel / mass, 
    F at the wheel is engine power * gear ratio / ( radius of the wheel * engine rotation rate), and 
    speed is is engine rotation rate * wheel radius / gear ratio, and 
    that all reduces to acceleration = engine power / speed / mass)
safe velocity = the velocity such that slowdown == acceleration at m2c engine power
max velocity = the velocity such that slowdown == acceleration
    (side note: the velocity equation here works out to 
    Cdair * V^3 + Crolling * V^2 - effective engine power  = 0, which 
    can be solved via the cubic formula https://math.vanderbilt.edu/schectex/courses/cubic/ but
    the math is exceedingly ugly. Not hard to program, just ugly.)
```

The net result is this:

Vehicle | Old Safe (mph) | Old Max (mph) |   | Coefficient Air | Air Area | Crolling | New Cruise (mph) | New Top (mph)
-- | -- | -- | -- | -- | -- | -- | -- | --
Electric Scooter (small motor) | 11 | 15 |   | 1 | 0.7 | 0.5 | 56 | 56
Electric Bike (very small motor) | 2 | 3 |   | 1 | 0.7 | 0.5 | 26 | 26
Electric Scooter | 35 | 48 |   | 1 | 0.7 | 0.5 | 93 | 97
Electric Bike (small motor) | 12 | 15 |   | 1 | 0.7 | 0.5 | 54 | 54
Scooter | 21 | 45 |   | 1 | 0.7 | 0.7 | 67 | 80
Motorcycle Sidecar | 32 | 77 |   | 1 | 1.7 | 1.944 | 60 | 71
Superbike | 91 | 194 |   | 1 | 0.7 | 1.4 | 109 | 130
Golf Cart (small motor) | 48 | 78 |   | 0.5 | 1.8 | 3.6 | 48 | 48
Golf Cart | 48 | 78 |   | 0.5 | 1.8 | 3.6 | 82 | 85
Hatchback | 64 | 192 |   | 0.5 | 2.7 | 17.1 | 77 | 93
Electric Car | 159 | 313 |   | 0.4 | 2.7 | 19.8 | 111 | 116
Solar Car | 58 | 125 |   | 0.3 | 3.6 | 15.1 | 82 | 85
Car | 63 | 192 |   | 0.45 | 2.7 | 23.1 | 75 | 91
Sports Car | 314 | 937 |   | 0.3 | 2.7 | 15.7 | 159 | 192
FBI Car | 167 | 504 |   | 0.5 | 3.06 | 25.7 | 101 | 123
Cube Van | 75 | 310 |   | 0.6 | 8.44 | 49.5 | 58 | 70
RV | 71 | 309 |   | 0.7 | 8.44 | 74.4 | 52 | 63
School Bus | 94 | 309 |   | 0.65 | 7.44 | 58.6 | 58 | 69
Luxury RV | 77 | 309 |   | 0.7 | 5.08 | 68.8 | 60 | 72
Flatbed Truck | 85 | 309 |   | 0.6 | 3.6 | 41.8 | 74 | 87
Armored Car | 138 | 504 |   | 0.6 | 4.08 | 47.4 | 85 | 101
Humvee | 138 | 504 |   | 0.57 | 4.08 | 45.0 | 87 | 103
Swat Truck | 76 | 309 |   | 0.6 | 8.44 | 98.9 | 52 | 62
APC | 70 | 309 |   | 0.6 | 7.92 | 91.9 | 54 | 64
Military Truck | 105 | 504 |   | 0.6 | 8.44 | 98 | 64 | 76
Semi Truck | 102 | 504 |   | 0.65 | 9.06 | 113.9 | 60 | 71
Tanks: Mobile Gun System | 69 | 309 |   | 0.65 | 8.58 | 107.9 | 50 | 60
Tanks: Atomic Mini Tank | 132 | 784 |   | 1 | 7.92 | 153.3 | 72 | 76
Tanks: Light Tank | 93 | 1153 |   | 1 | 8.58 | 166.0 | 72 | 86
Tanks: Main Battle Tank | 60 | 1262 |   | 1 | 12.21 | 236.3 | 65 | 77

From here, we can adjust some of the numbers or vehicle designs: the electric scooter could drop to using a 10 HP small electric motor, instead of a 50 HP electric motor, and putter along at 50+ mph instead of maxing out at 15 mph. With 1/5 the power draw, the battery life of the electric scooter would increase by 5x. Similarly with the golf cart and electric bicycles.

In general, this change would regularize vehicle speed: small vehicles with small or moderate engines generally go faster, heavy vehicles with large engines are a little slower but almost all keep reasonable cruise speeds (only the RV, SWAT trucks, and APC can't make 55 mph at cruise speeds) while drastically limiting the crazy top speeds that CDDA is notorious for.

**Describe alternatives you've considered**  
A more detailed model could take account of gear ratios and transmission losses, but this is good enough for now.

**Additional context**  
See https://docs.google.com/spreadsheets/d/1jPrZc4Tzl7S6rp63ek_09L1mB1wAVSyYtKORyUMyhCw/edit#gid=0 for the tables of values I used for this analysis.

### PR DESCRIPTION
**Pull request by mlangsdorf**
#### Summary
```SUMMARY: Balance "Rework vehicle safe and max velocities based on physics"```

#### Purpose of change
Fixes #25652 

CDDA is notorious for allowing people to build vehicles with ridiculous top speeds.  It is also makes it
hard to create realistic, small vehicles with small engines: speeds are tied directly to engine power, and vehicle mass makes very little difference.  As such, CDDA ends up with main battle tanks that can
break the sound barrier and bicycles that need 25 HP motors just to move at reasonable speeds.

The solution is to rework vehicle speeds based on physics.  Broadly speaking, an engine delivers a
certain amount of force to the wheel to accelerate the vehicle, and that wheel force is resisted by drag force.  Drag force is the sum of the air drag force, which varies with the square of vehicle speed and is proportional  to the cross sectional area of the vehicle times a constant value that depends on vehicle streamlining, and the rolling resistance, part of which is constant and part of which varies with vehicle speed and both are proportional to vehicle mass times a constant value based on wheel design. A vehicle reaches maximum speed when wheel force is exactly equal to drag force: the net forces are 0 and the vehicle can't accelerate any more.

As usual in physics, acceleration is net force divided by mass.

#### Describe the solution
Introduce two new variables: coeff_air_drag and coeff_rolling_resistance.

* coeff_air_drag is calculating by raycasting down the vehicle's columns to determine the vehicle's aerodynamic profile, then multiplying the results by the vehicle's width and approximated height. Vehicle height is increased for having aisles, fullboards, roofs, solar panels, etc. 
* coeff_rolling_drag is calculating by summing the rolling resistance values for each wheel and multiplying the result by the vehicle's weight.

coeff_air_drag is roughly a replacement for the existing k_aerodynamics, and coeff_rolling_drag is roughly a replacement for the existing k_friction.

wheel force is calculated as engine power / speed (this abstracts a lot of gearing but it's a workable assumption).
air drag force is calculated as coeff_air_drag * speed^2.
rolling resistance is calculated as coeff_rolling_drag * ( constant + speed ).

Once each turn, in vehicle::gain_moves(), calculate slowdown as sum of air drag force and rolling resistance divided by vehicle mass, then later apply acceleration at the current speed as current wheel force divided by mass, up to the limit of set cruise speed.

The net effect: vehicles tend to accelerate a little faster at low speeds and slow down much more quickly at high speeds.  Light weight vehicles with low HP engines can be fast.

#### Describe alternatives you've considered
Calculating air drag is expensive but I don't see a good way to keep the improved physical simulation and make it faster.

Some vehicles and vehicle parts will need to be adjusted.  For instance, now that 2 HP motors produce realistic amounts of power for lightweight vehicles, foot cranks don't need to produce 16 HP.  A couple of vehicles such as the superbike could use some minor adjustments to make them more streamlined.

#### Additional context
This is still a work in progress, but is hopefully working well enough for limited testing.  One obvious bug is that coeff_air_drag and coeff_rolling_drag are cached when calculated, but currently there's no code to dirty the caches.

Physics, math, and derivation of most values are at #25652.  A chart of expected vehicle speeds for most vehicles in the game is here: https://docs.google.com/spreadsheets/d/1jPrZc4Tzl7S6rp63ek_09L1mB1wAVSyYtKORyUMyhCw/edit#gid=81854995

My sincerest and deepest thanks to @cake-pie, who did all the hard work in correcting my half-assed formulas and then digging out obscure physical constants.  Coding is easy, research is hard.

## ALERT
Fuel consumption has increased dramatically.  On the vehicle efficiency tests, most vehicles are getting about 1/6th to 1/3rd the range they used to get on the constant speed pavement and dirt tests (and vehicles are having to go much slower on dirt to achieve that speed).  The road stop and dirt stop numbers are more favorable, with some vehicles getting a 33% to 80% range boost and no vehicle doing worse than halving the range.  The one exception to the general fall-off are scooters which now get a 60% range boost at constant speed and a 2000% boost on the start/stop tests.

### CODE REVIEW

**Comment by Szara-ManOfHonor** [comment]
Could the performance load of air drag be reduced by only recalculating when a part is added to or removed from the vehicle?

**Comment by mlangsdorf** [comment]
Yeah, the intent is only to dirty the cache when a part is added on removed.  The problem is that during a vehicle collision, a lot of parts might get removed very rapidly, and I'd like to avoid repeatedly recalculating coeff_air_drag when the results are going to discarded as soon as the next part is broken off.

**Comment by mlangsdorf** [comment]
Dropping foot cranks down to a workable 600W (150-250W would be a reasonable value, but upping it for game purposes) means that electric bikes can't use any of the current alternators.  Though apparently bicycle alternators are usually less than 1/2HP so that's not crazy.

**Comment by Fris0uman** [comment]
Foot crank + truck alternator is currently a way to get infinite electricity since you can pedal and craft at the same time, does that mean that it won't be possible anymore ? Also does that mean that IRL it's not possible to manually turn a car or truck alternator ?

**Comment by mlangsdorf** [comment]
Currently, foot cranks produce 12 HP, + 2.5 HP per point of Strength.  Truck alternators require 3 HP of engine power.

Realistically, foot cranks should produce <1 HP for normal humans. It shouldn't be possible for a normal human to drive a truck, car, or motorcycle alternator.

Infinite electrical power would be possible from a new 0.2 HP bike alternator, but it would take a lot of time because that's only 150W of power generation.

**Comment by Fris0uman** [comment]
So it seems pretty good that foot crank won't be compatible with those alternators then. It just means we need to add a bicycle alternator, or are they already in ?

**Comment by mlangsdorf** [comment]
Need to add a bike alternator.
Possibly add some code to convert all alternators on a foot crank into bike alternators.

**Comment by Szara-ManOfHonor** [comment]
For the post crash recalculation, would it be unreasonable to add a six (game) second grace period on recalculations?
So in steps it plays out like this
Crash!
Part torn off
Recalculation
Effect grace period added with duration six seconds (add to what I don't know)
Additional parts torn off, no effect because of grace period
Effect grace period terminates and recalculates the coefficient.

Hmn, but that only works *well* if vehicles can have an effect. Or maybe not. Maybe other ways. If vehicles can have a variable that is used for 'need recalc', then when the grace period ends, all vehicles in the bubble can be checked for that variable?

**Comment by mlangsdorf** [comment]
I think you're describing a two-stage cache control system, which is what I ended up using.
When you add a vehicle part, remove a vehicle part, or break a vehicle part, the coeff_air_changed variable is set. in part_removal_cleanup(), if coeff_air_changed is set, coeff_air_dirty is set and coeff_air_changed is cleared.  The next time someone tries to retrieve coeff_air_dirty(), it's recalculated, cached, and coeff_air_dirty is cleared.

If you're driving down the road when a mininuke explodes, shattering your RV so that it splits off 3-4 new vehicles, the original RV hulk keeps its coeff_air_drag until after the parts are cleaned up at the end of vehicle movement.

**Comment by kevingranade** [comment]
I don't see any blockers for merging this, any particular concerns about things I should check?

I did notice it's spamming some debug messages, that needs to get cleaned up.

The efficiency is fine.
I did some spot checks of acceleration and top speed, and I didn't see anything out of line.

One goofy regression is if you start from a halt, set speed to 10MPH, and then "turn" multiple times in a row, you can end up pivoting 90deg at a time.

**Comment by mlangsdorf** [comment]
Can't reproduce the regression.  Starting from from start on a car, a scooter, a humvee, and a schoolbus, each one skewed as it turned at 10 mph.  Tried it for both left and right hand turns.  Is there more to the regression reproduction?

**Comment by ZhilkinSerg** [comment]
I couldn't reproduce it either.

**Comment by kevingranade** [comment]
More info on the turning thing.
Vehicle heading indicator enabled.
Spawn in an ordinary car, on dirt facing north.
Board, start engine.
Set speed to 10.

1. Hit "RIGHT", vehicle heading indicator rotates, speed doesn't change.
2. Hit "RIGHT", vehicle heading indicator rotates, reported speed goes to 5.
3. Hit "RIGHT", vehicle heading indicator rotates (~45deg from ), reported speed goes to 10.
4. Hit "RIGHT", vehicle heading indicator rotates (a bit more, still ~45deg ish), reported speed at 10, vehicle finally moves and snaps to 45deg rotation.

At step 3:
2018-12-04-041129_873x420_scrot

At step 4:
2018-12-04-041141_851x406_scrot

**Comment by mlangsdorf** [comment]
I can reproduce that on an unmodified game by setting all the grass to long grass, and then driving and turning right.  Sometimes - not consistently, but sometimes - the first displacement will go to 45 degrees, though going to 15 or 30 is much more common.

However, it does happen 100% of the time with the new code, so it's kind of a regression, and you can also spin a car by stopping it, turning it 180 degrees, and only then applying gentle acceleration.

it looks like I wasn't setting turn_dir = face.dir in thrust() when calling the first stop(), but I also suspect that moving that assignment into stop() might be the best solution.

File: tests/vehicle_drag.cpp [hunk 1]
```diff
@@ -0,0 +1,288 @@
...
+    valid &= d_in_bounds( expected_c_rr, c_rolling );
+    valid &= i_in_bounds( expected_safe, safe_v );
+    valid &= i_in_bounds( expected_max, max_v );
+    return valid;
+}
+
+// Behold: power of laziness
+void print_drag_test_strings( const std::string &type )
+{
+    std::ostringstream ss;
```
Comment (ZhilkinSerg) [inline tests/vehicle_drag.cpp]: image

Comment (mlangsdorf) [reply tests/vehicle_drag.cpp]: Drat.  Will fix.

Comment (mlangsdorf) [reply tests/vehicle_drag.cpp]: Man, I don't even know.

File: tests/vehicle_drag.cpp [hunk 2]
```diff
@@ -0,0 +1,289 @@
...
+    if( veh_ptr == nullptr ) {
+        return;
+    }
+    ss << "    test_vehicle_drag( \"" << type << "\", ";
+    ss << veh_ptr->coeff_air_drag() << ", ";
+    ss << veh_ptr->coeff_rolling_drag() << ", ";
+    ss << veh_ptr->safe_velocity() << ", ";
+    ss << veh_ptr->max_velocity();
+    ss << " );" << std::endl;
+    printf( "%s", ss.str().c_str() );
```
Comment (BevapDin) [inline tests/vehicle_drag.cpp]: Have you considered printing this directly to `std::cout`? Why the separate string stream?

**Pull request merged**